File: Python/14.Longest Common Prefix.py
#找字符串数组的最长共同前缀子串
class Solution(object):
    def longestCommonPrefix(self, strs):
        if len(strs)<=1:
            return strs[0] if len(strs)==1 else ""
        minlen = min([len(st) for st in strs])
        i = minlen
        while i>0:
            newstrs = [strs[j][0:i] for j in range(len(strs))]
            trueorfalse = [newstrs[k]==newstrs[k+1] for k in range(len(newstrs)-1)]
            if False in trueorfalse:
                i-=1
                continue
            else:
                return newstrs[0]
        return ""

s = Solution()
#print(s.longestCommonPrefix(['abcd','abc','abceee','abcfef']))
print(s.longestCommonPrefix(['a','b']))

File: C++/String/14.Longest Common Prefix.cpp
/*
寻找给出的所有string的最长共同前缀子串。
*/
/*
思路：对第一个字符串的每一个字母循环找即可。
*/
#include <iostream>
#include <cstring>
#include <vector>
using namespace std;

class Solution {
public:
	string longestCommonPrefix(vector<string>& strs) {
        int strlength = strs.size();
        if(strlength==0)	return "";
        string result = "";
        for(int i=0; i < strs[0].length（）; ++i)
        {
        	int ch = strs[0][i];	
        	for(int j=1; j < strlength; ++j)
        	{
        		if(i>=strs[j].length() || strs[j][i]!=ch)
        			return result;
        	}
        	result += ch;
        }
        return result;
    }
};

int main()
{
	Solution s;;
	vector<string> strs{"abc","abced","abde"};
	cout<<s.longestCommonPrefix(strs)<<endl;
}

